api/oc_push.c

```diff
@@ -0,0 +1,2658 @@
+ *
+ * 			each list member is instance of `oc_recvs_t`
+ */
+OC_LIST(recvs_list);
+
+
+
+/*
+ * @brief	memory block definition for storing properties representation of pushed resource
+ */
+OC_MEMB(rep_instance_memb, oc_rep_t, 1);
+
+/*
+ * @brief	memory block definition for storing pushed resource representation list
+ */
+OC_MEMB(pushd_rsc_rep_instance_memb, oc_pushd_rsc_rep_t, 1);
+
+
+
+/*
+ * @brief	`pushed_rsc_list` keeps Resource representation of Pushed Resources
+ */
+OC_LIST(pushd_rsc_rep_list);
+
+
+
+/*
+ * @brief	process which handles push notification
+ */
+OC_PROCESS(oc_push_process, "Push Notification handler");
+
+
+
+char *pp_state_strs[] =
+{
+		"waitingforprovisioning",			/*OC_PP_WFP*/
+		"waitingforupdate",					/*OC_PP_WFU*/
+		"waitingforresponse",				/*OC_PP_WFR*/
+		"waitingforupdatemitigation",		/*OC_PP_WFUM*/
+		"waitingforresponsemitigation",	/*OC_PP_WFRM*/
+		"error",									/*OC_PP_ERR*/
+		"timeout"								/*OC_PP_TOUT*/
+};
+
+
+
+char *cli_status_strs[] =
+{
+		"OC_STATUS_OK",                        /* 0 */
+		"OC_STATUS_CREATED",                   /* 1 */
+		"OC_STATUS_CHANGED",                   /* 2 */
+		"OC_STATUS_DELETED",                   /* 3 */
+		"OC_STATUS_NOT_MODIFIED",              /* 4 */
+		"OC_STATUS_BAD_REQUEST",               /* 5 */
+		"OC_STATUS_UNAUTHORIZED",              /* 6 */
+		"OC_STATUS_BAD_OPTION",                /* 7 */
+		"OC_STATUS_FORBIDDEN",                 /* 8 */
+		"OC_STATUS_NOT_FOUND",                 /* 9 */
+		"OC_STATUS_METHOD_NOT_ALLOWED",        /* 10 */
+		"OC_STATUS_NOT_ACCEPTABLE",            /* 11 */
+		"OC_STATUS_REQUEST_ENTITY_TOO_LARGE",  /* 12 */
+		"OC_STATUS_UNSUPPORTED_MEDIA_TYPE",    /* 13 */
+		"OC_STATUS_INTERNAL_SERVER_ERROR",     /* 14 */
+		"OC_STATUS_NOT_IMPLEMENTED",           /* 15 */
+		"OC_STATUS_BAD_GATEWAY",               /* 16 */
+		"OC_STATUS_SERVICE_UNAVAILABLE",       /* 17 */
+		"OC_STATUS_GATEWAY_TIMEOUT",           /* 18 */
+		"OC_STATUS_PROXYING_NOT_SUPPORTED"     /* 19 */
+};
+
+
+/*
+ * mandatory property of oic.r.pushporxy, oic.r.pushreceivers
+ */
+enum {
+	PP_PUSHTARGET = 0x01,
+	PP_SOURCERT = 0x02,
+	PR_RECEIVERS = 0x01,
+	PR_RECEIVERURI = 0x02,
+	PR_RTS = 0x04
+};
+
+
+/*
+ * if this callback function is provided by user, it will called whenever new push is arrived...
+ */
+void (*oc_push_arrived)(oc_pushd_rsc_rep_t *) = NULL;
+
+
+#define pp_statestr(i) (pp_state_strs[(i)])
+
+/*
+ * @brief update Push Proxy state from state to new_state
+ *
+ * @param state		oc_string_t
+ * @param new_state	char *
+ */
+#define pp_update_state(state, new_state) \
+{ \
+	oc_free_string(&(state)); \
+	oc_new_string(&(state), (new_state), strlen((new_state))); \
+}
+
+/*
+ * @brief initialize oc_string_t object
+ */
+#define oc_init_string(str) \
+{ \
+	(str).size = 0; \
+	(str).ptr = NULL; \
+	(str).next = NULL; \
+}
+
+
+
+static oc_interface_mask_t _get_ifmask_from_ifstr(char *ifstr)
```

Comment:
Already existing `oc_ri_get_interface_mask` can be used instead of this new function